YuJeongMin/YuJeongMin211018/CameraControl.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using UniRx;
+using UniRx.Triggers;
+using UnityEngine;
+
+namespace InGame
+{
+    public class CameraControl : MonoBehaviour
+    {
+        private Camera _camera;
+
+        private const float ScreenRatio = 9f / 16f;
+        private const float DefaultSize = 5f;
+        private const float DefaultPosX = -10;
+        private const float Half = 0.5f;
+
+        [SerializeField] private Transform myCharacterTransform;
+        [SerializeField] private Transform anotherCharacterTransform;
+
+        public void SetTarget(Transform my, Transform another)
+        {
+            myCharacterTransform = my;
+            anotherCharacterTransform = another;
+        }
+
+        public void SetTarget(GameObject my, GameObject another)
+        {
+            myCharacterTransform = my.transform;
+            anotherCharacterTransform = another.transform;
+        }
+
+        private void Start()
+        {
+            _camera = GetComponent<Camera>();
+
+            this.UpdateAsObservable().Subscribe(_ =>
+            {
+                var my = myCharacterTransform.position;
+                var another = anotherCharacterTransform.position;
+                
+                transform.position = CameraPosition(my, another);
+                _camera.orthographicSize = CameraSize(my, another);
+            });
```

Comment:
```suggestion
            }).AddTo(gameObject);
```
`UpdateAsObservable` 과 같은 스트림을 만드는 행위의 경우, `Subscribe` 메소드에서 보통 `IDisposable` 객체를 반환하게 됩니다. 이는 해당 스트림을 종료시키기 위해서인데, 종료시켜주지 않을 경우 이 `UpdateAsObservable` 을 시작시킨 오브젝트가 사라지더라도 해당 스트림은 UniRx 스레드에서 돌고있게 됩니다.

이를 방지하기 위해 반환되는 `IDisposable` 객체를 필드로 저장해뒀다가 해당 객체 또는 스트림이 필요없어지는 시점에 개발자가 직접 `Dispose()` 를 해주거나, `AddTo` 메소드를 이용해 특정 오브젝트에 컴포넌트로 붙여서 해당 오브젝트가 사라지는 시점에 스트림도 함께 GC에서 처리하도록 하는 방법이 있습니다.

Reply:
조언 감사합니다. 그런데 검색에서 잘 안 나오는 부분이 있어 여쭤볼 게 있습니다. 유니티에서 생명주기에 따라 스트림이 계속 돌고 있는 오류가 UpdateAsObservable과 같은 스트림에서 나타난다면 어떤 스트림에서 일어나는지에 대해 더 자세히 알고 싶습니다.

혹시 작성한 코드에서 this 포인터에서 가져오는 스트림의 경우 그런 건가요?

Reply:
요지는 `UpdateAsObservable` 이 문제가 아니라, 기본적으로 모든 스트림을 만드는 행위는 해당 스트림을 UniRx 스레드에서 처리하게 되고 이를 유니티 메인 스레드에서 동기화하며 동작하게 됩니다.

여기서 중요한건 스트림을 만든 객체 내에서 저장된 채로 해당 객체와 함께 동작하는게 아니라, UniRx 스레드를 통해 개별적으로 돌게 된다는 것입니다. 예시를 들어보겠습니다.

하나의 클래스를 만들어서 모노비헤이비어를 상속시키고, 이 객체 안에서 종류에 관계없이 스트림을 하나 만들어 멤버 필드로 만들어놓았습니다.
여기서 스트림을 필드로 갖고있는 객체를 A, A 내에서 생성된 스트림을 B라고 하겠습니다.

여기서 A는 B를 멤버로 갖고있긴 하나, B의 스트림과 관련된 동작 자체는 A 내부에서 하지 않고 생성과 동시에 UniRx 스레드에서 관리를 하게 됩니다. 즉, A의 존재 여부와는 관계없이 UniRx 스레드가 관리하고 있으므로 A가 사라져도 B는 사라지지 않는다는 뜻입니다.

당연합니다. A가 스트림을 관리하는게 아니라, 그 참조를 필드의 형태로 갖고있을 뿐이니까요.
애시당초 스트림은 UniRx 스레드가 관리합니다. 어디서 만들었는지는 그닥 중요하지 않고요.
이 때문에, A가 파괴되건 말건 B 스트림은 계속 살아있고, 데이터를 큐에 계속 받게 됩니다.

이러한 스트림을 객체의 라이프 사이클과 함께 처리해주기 위해 AddTo 라는 메소드를 이용하여 스트림을 컴포넌트의 형태로 객체에게 종속시키거나, `IDisposable` 객체를 받아와서 필요한 시점에 같이 Dispose 해주는 것입니다.

요약하겠습니다.
```
1. UniRx 스트림은 어떤 객체에서 만들었건간에 UniRx 스레드에서 관리한다.
2. 따라서 스트림은 어떤 객체에서 만들었건간에 해당 게임오브젝트가 사라지던가 말던가 꿋꿋이 계속 돈다. 객체와는 아무 관계가 없다. 스트림의 종류 자체에도 관계가 없다.
3. 이러면 스트림의 메모리 관리가 어렵기 때문에, AddTo를 이용해 컴포넌트 형태로 객체에 종속시키거나 IDisposable 객체를 필드로 저장했다가 개발자가 직접 필요한 시점에 Dispose를 해주어야 한다.
```